### PR TITLE
Android ZOrder for transparent surfaceView

### PR DESCRIPTION
**Pull request by CedricGuillemet**
As discussed offline, add a new `androidView` property and remove `isTopMost`

### CODE REVIEW

**Comment by bghgary** [comment]
I couldn't find any info on this. What is the difference between these two functions?

**Comment by CedricGuillemet** [comment]
> I couldn't find any more info on this. What is the difference between these two functions in practice?

`setzorderontop` will make the surfaceView on top of everything that's under (surfaceView, textureView,... anything), independently of view hierarchy.
`setzordermediaoverlay` will composite the surfaceView on top of other surfaceViews

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java [hunk 1]
```diff
@@ -108,9 +108,9 @@ private void setIsTransparentAndIsTopMost(Boolean isTransparent, Boolean isTopMo
                 surfaceHolder.setFormat(PixelFormat.TRANSPARENT);
             }
             if (isTopMost) {
-                // ZOrder is not dynamic before Android 11. Recreate the surfaceView and set order before adding to the parent
-                // https://developer.android.com/reference/android/view/SurfaceView#setZOrderOnTop(boolean)
-                this.surfaceView.setZOrderOnTop(true);
+                // setZOrderMediaOverlay  is not dynamic before Android 11. Recreate the surfaceView and set order before adding to the parent
```
Comment (ryantrem) [inline Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:111]:
```suggestion
                // setZOrderMediaOverlay is not dynamic before Android 11. Recreate the surfaceView and set order before adding to the parent
```

**Comment by ryantrem** [comment]
`setZOrderMediaOverlay` says *"the surface view's surface is placed on top of another regular surface view in the window (but still behind the window itself)"*. Doesn't that mean it is not top most (i.e. it is still behind the window itself and therefore behind all the normal UI)?

**Comment by CedricGuillemet** [comment]
> `setZOrderMediaOverlay` says _"the surface view's surface is placed on top of another regular surface view in the window (but still behind the window itself)"_. Doesn't that mean it is not top most (i.e. it is still behind the window itself and therefore behind all the normal UI)?

Yes, it's not topmost on Android anymore (as it's still on iOS). It's topMost on other surfaceViews. It's platform specific behaviors. I don't see any obvious name for that. Or even if just a name change is enough.

**Comment by bghgary** [comment]
It seems like this flag is no longer consistent across platforms with this change. That seems bad. I wonder if it makes sense to expose the underlying platform view and the application can do whatever it wants with it?

**Comment by CedricGuillemet** [comment]
> It seems like this flag is no longer consistent across platforms with this change. That seems bad. I wonder if it makes sense to expose the underlying platform view and the application can do whatever it wants with it?

Do you mean exposing a MTKView or a surfaceView to JS? Or asking user to add a bridge between the native view and JS?

**Comment by bghgary** [comment]
> Do you mean exposing a MTKView or a surfaceView to JS? Or asking user to add a bridge between the native view and JS?

I'm not sure. Is there already an existing JS visible interface for talking to the native view? I'm guessing no.

**Comment by ryantrem** [comment]
What I've seen in other RN components is having props that are platform specific and documented as such. In this specific case, what is the behavior you are trying to achieve? It seems like maybe there is a SurfaceView, and then the BRN EngineView is supposed to go on top of that, but then UI is supposed to go on top of everything? Although that's not the behavior we'd have on iOS, right? So yea, maybe if you can explain a little more scenario that will help with figuring out a reasonable set of props across platforms.

**Comment by CedricGuillemet** [comment]
The scenario is quite simple actually: when isTransparent is true for the EngineView, there is nothing to do on iOS or Windows except.
On Android, there are 3 modes: textureView, surfaceView with zTop, surfaceView with with mediaOverlay.
So, we can removed `isTopMost` and have an Android enum property 'androidTransparencyMode`.
Or keep `isTopMost` and add another property that will interfer.
Is it too late to remove isTopMost?

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java [hunk 2]
```diff
@@ -94,30 +95,37 @@ private void setIsTransparentAndIsTopMost(Boolean isTransparent, Boolean isTopMo
...
                 surfaceHolder.setFormat(PixelFormat.TRANSPARENT);
             }
-            if (isTopMost) {
-                // ZOrder is not dynamic before Android 11. Recreate the surfaceView and set order before adding to the parent
-                // https://developer.android.com/reference/android/view/SurfaceView#setZOrderOnTop(boolean)
+            if ((viewMode.equals("") && isTransparent) || viewMode.equals("SurfaceViewZTopMost"))
+            {
                 this.surfaceView.setZOrderOnTop(true);
+            } else if (viewMode.equals("SurfaceViewZMediaOverlay"))
+            {
```
Comment (bghgary) [inline Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:119]:
```suggestion
            } else if (viewMode.equals("SurfaceViewZMediaOverlay")) {
```

**Comment by bghgary** [comment]
It would be good if @ryantrem can review as well before I sign off.

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java [hunk 3]
```diff
@@ -94,30 +95,37 @@ private void setIsTransparentAndIsTopMost(Boolean isTransparent, Boolean isTopMo
             this.transparentTextureView.setVisibility(View.GONE);
             this.transparentTextureView = null;
         }
-        if (isTransparent && !isTopMost) {
+
+        //if (isTransparent) {
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:99]: Fixed.

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java (hunk 2 above)
Comment (CedricGuillemet) [reply Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:119]: Fixed

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java [hunk 4]
```diff
@@ -33,14 +33,14 @@ public final class EngineView extends FrameLayout implements SurfaceHolder.Callb
 
     private SurfaceView xrSurfaceView;
     private boolean isTransparent = false;
-    private boolean isTopMost = false;
+    private String viewMode = "";
```
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:36]: Fixed

File: Modules/@babylonjs/react-native/README.md [hunk 5]
```diff
@@ -106,3 +98,19 @@ e.g.
 ```
 
 Note: Currently only one `EngineView` can be active at any given time. Multi-view will be supported in a future release.
+
+On Android, specific `androidView` property can help set the type of the view used for rendering. Depending on user needs and performance, refer to this table:
```
Comment (ryantrem) [inline Modules/@babylonjs/react-native/README.md:102]: Maybe
```suggestion
The Android specific `androidView` property can help set the type of the view used for rendering. Depending on user needs and performance, refer to the table below. [`TextureView`](https://developer.android.com/reference/android/view/TextureView) can be inserted anywhere in the view hierarchy, but is less efficient. [`SurfaceView`](https://developer.android.com/reference/android/view/SurfaceView) can only be full above or fully below the rest of the UI, but is more efficient.
```

File: Modules/@babylonjs/react-native/README.md [hunk 6]
```diff
@@ -106,3 +98,19 @@ e.g.
...
+| Transparent | androidView value        | Description |
+| ----------- | ------------------------ | ----------- |
+| False       | TextureView              | Opaque TextureView
+| False       | SurfaceView              | Simple surfaceView (default when no `androidView` set with Transparency Off)
+| False       | SurfaceViewZTopMost      | SurfaceView with ZTopMost set to true. https://developer.android.com/reference/android/view/SurfaceView#setZOrderOnTop(boolean)
+| False       | SurfaceViewZMediaOverlay | SurfaceView with ZMediaOverlay set to true. https://developer.android.com/reference/android/view/SurfaceView#setZOrderMediaOverlay(boolean)
+| True        | TextureView              | Transparent TextureView, less performant than transparent SurfaceViews
+| True        | SurfaceView              | SurfaceView will stay opaque
+| True        | SurfaceViewZTopMost      | SurfaceView with ZTopMost set to true. (default when no `androidView` set with Transparency On) Transparent but top most.
+| True        | SurfaceViewZMediaOverlay | SurfaceView with ZMediaOverlay set to true. Only Transparent ontop of other SurfaceViews
```
Comment (ryantrem) [inline Modules/@babylonjs/react-native/README.md:113]:
```suggestion
| isTransparent | androidView        | Description |
| ----------- | ------------------------ | ----------- |
| False       | TextureView              | Opaque TextureView.
| False       | SurfaceView              | Simple surfaceView (default when no `androidView` set with `isTransparent=false`).
| False       | SurfaceViewZTopMost      | SurfaceView with [ZTopMost](https://developer.android.com/reference/android/view/SurfaceView#setZOrderOnTop(boolean)) set to `true`.
| False       | SurfaceViewZMediaOverlay | SurfaceView with [ZMediaOverlay](https://developer.android.com/reference/android/view/SurfaceView#setZOrderMediaOverlay(boolean)) set to `true`.
| True        | TextureView              | Transparent TextureView.
| True        | SurfaceView              | SurfaceView will stay opaque
| True        | SurfaceViewZTopMost      | SurfaceView with [ZTopMost](https://developer.android.com/reference/android/view/SurfaceView#setZOrderOnTop(boolean)) set to `true`. Transparent but top most. (default when no `androidView` set with `isTransparent=true`)
| True        | SurfaceViewZMediaOverlay | SurfaceView with [ZMediaOverlay](https://developer.android.com/reference/android/view/SurfaceView#setZOrderMediaOverlay(boolean)) set to `true`. Only Transparent on top of other SurfaceViews.
```

File: Modules/@babylonjs/react-native/EngineView.tsx [hunk 7]
```diff
@@ -116,7 +116,7 @@ export const EngineView: FunctionComponent<EngineViewProps> = (props: EngineView
     if (initialized !== false) {
         return (
             <View style={[{ flex: 1 }, props.style, { overflow: "hidden" }]}>
-                { initialized && <NativeEngineView ref={engineViewRef} style={{ flex: 1 }} onSnapshotDataReturned={snapshotDataReturnedHandler} isTransparent={isTransparent} antiAliasing={antiAliasing} isTopMost={isTopMost}/> }
+                { initialized && <NativeEngineView ref={engineViewRef} style={{ flex: 1 }} onSnapshotDataReturned={snapshotDataReturnedHandler} isTransparent={isTransparent} antiAliasing={antiAliasing} androidView={androidView}/> }
```
Comment (ryantrem) [inline Modules/@babylonjs/react-native/EngineView.tsx:119]: If the `androidView` property is set on a control that doesn't support the property (e.g. the underlying component for all platforms other than Android), does it no produce an error or warning or anything?

Comment (CedricGuillemet) [reply Modules/@babylonjs/react-native/EngineView.tsx:119]: I didn't see any error or warning.

**Review by bghgary — APPROVED** [review]
Good from my perspective.

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ryantrem) [reply Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java:99]: I assume the "remove" comment was about the commented out `//if (isTransparent) {`, which appears to still be there.

**Review by ryantrem — APPROVED** [review]

File: Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (CedricGuillemet) [reply Modules/@babylonjs/react-native-iosandroid/android/src/main/java/com/babylonreactnative/EngineView.java]: Indeed! idk what happened. Gone now.

**Pull request merged**